meshmode/dof_array.py

```diff
@@ -170,6 +169,24 @@ def __str__(self):
     def __repr__(self):
         return f"DOFArray({repr(self._data)})"
 
+    # {{{ array container protocol
+
+    def as_iterable(self):
+        return ((i, subary) for i, subary in enumerate(self._data))
+
+    @classmethod
+    def from_iterable(cls, actx, iterable):
+        iterable = list(iterable)
+
+        result = [None] * len(iterable)
+        for i, subary in iterable:
+            result[i] = subary
+
+        assert all(subary is not None for subary in result)
+        return DOFArray(actx, tuple(result))
```

Comment:
```suggestion
    return DOFArray(actx, data=tuple(subary for _, subary in iterable))
```
? (untested but just a thought)

Reply:
The original code was supposed to do a bit of checking that it's not being passed nonsense. I'll make that a bit more explicit, but I don't think I'd like getting rid of it completely.

This interface isn't particularly meant to be used by the outside world, so I could be convinced to get rid of it?